Addressing Issue #43: added csv localisation support

Deccir opened the pull request:
- added godots localization to automatically translate text, speaker, options and string variables of a dialogue based on a .csv giving the user the option to either use separate .tres for each language or have one single .tres that is then translated using a .csv.
- added an example

nagidev committed:
Initial commit

nagidev committed:
First commit

nagidev committed:
Added load tree functionality

Trying to open test.json using Load Tree now works

nagidev committed:
Bugfix #1 + Added Examples

1. Loading a tree while a tree is already open works properly now.
2. Added example files to teach the basics of making dialogues.
3. The dialogues in the demo now animate using Tweens.

nagidev committed:
Demo dialogues can be skiiped now

nagidev committed:
Comment nodes can be saved now

nagidev committed:
Switched from bottom panel to main screen

nagidev committed:
Bugfix #2 + Shortcut support

1. Ctrl + S - saves the currently open tree
2. Added a side panel with shortcuts to add nodes to graph quickly
3. Dialogue node connections now disconnect on removing options
4. Ctrl + D duplicates nodes (Bugfixes required)

nagidev committed:
Bugfix #3

1. Duplicate (Ctrl + D) works with single node selected.
2. Fixed slots not disappearing after removing options.

nagidev committed:
Bugfix #4 + Save Indicator + Icon

nagidev committed:
Fixed Example1.json

nagidev committed:
Bugfix #5

Comment nodes are not apended twice in commentNodes array now.

nagidev committed:
Multiple Tree Support + Bugfix #6

Multiple trees can be made in a single file, each with a unique ID. However, Quick Run & Run This Tree now only work when one of the Start nodes is selected.
Code to duplicate was rewritten, Node.duplicate() was buggy.

nagidev committed:
Code refactor #1

nagidev committed:
total refactor

nagidev committed:
Added SignalNode

nagidev committed:
updated demo scene

nagidev committed:
added screenshots + fixed example4.json

nagidev committed:
Revert "added screenshots + fixed example4.json"

This reverts commit 9ebb22131be5cd954c1edaa568c17c2c9e7d3f47.

nagidev committed:
Fixed example4.json + added screenshots

nagidev committed:
Update README.md

nagidev committed:
added gdignore file

nagidev committed:
Merge branch 'master' of https://github.com/Bhgt/DialogueNodes

nagidev committed:
Added Variable Set/Get functionality

nagidev committed:
Added Condition functionality

nagidev committed:
Updated version number

nagidev committed:
fixed dialogue node size issue, added file popup menu

nagidev committed:
v0.2: added dialogue text animation, improved bbcode support

nagidev committed:
Bugfix + Added example themes

nagidev committed:
Added skip function to bbcodeWait.gd

nagidev committed:
Bugfix + Added character_displayed signal

nagidev committed:
Ported to Godot 4.1

nagidev committed:
Updated README.md

Commit:
Disabled scroll follow.

nagidev committed:
Switched from JSON to Resources + Added Character Resource

nagidev committed:
Merge pull request #10 from CantyCanadian/main_DisableScrollFollow

Disabled scroll follow in DialogueBox.

nagidev committed:
Added bbcode ghost effect from Godot docs

nagidev committed:
Updated screenshots for Godot 4.1

nagidev committed:
Fixed import files for icons

nagidev committed:
Added option filtering + some QoL features

nagidev committed:
Updated variables section

nagidev committed:
Updated README.md

nagidev committed:
Bugfixes for DialogueBox

nagidev committed:
Updated README.md

nagidev committed:
Bugfixes for popup menus

nagidev committed:
Added better dialogue scrolling and documentation

nagidev committed:
Added some QoL fixes

nagidev committed:
Minor bug fixes

nagidev committed:
Fixed icon theming and scaling issues

nagidev committed:
Updated to version 1.0.2

nagidev committed:
Added line breaking in options

nagidev committed:
Updated Character.svg and created CharacterList.svg

nagidev committed:
Updated addon for Godot 4.2

nagidev committed:
Updated plugin version to 1.0.2a

nagidev committed:
Fixed memory leaks in workspace.gd

nagidev committed:
Fixed memory leaks in workspace.gd

nagidev committed:
Added option_selected signal to DialogueBox

nagidev committed:
Merge branch 'godot-4.2'

nagidev committed:
Updated README.md

nagidev committed:
Added bold and italic fonts for editor dialogue box

nagidev committed:
Fixed minor bug caused by hide_portrait

nagidev committed:
Removed unused enums in setNode.gd

nagidev committed:
Updated README.md

nagidev committed:
Added dialogue audio example

aaronfranke committed:
Add .gitattributes file for the Godot Asset Library

nagidev committed:
Update .gitattributes

nagidev committed:
Merge pull request #34 from aaronfranke/gitattr

Add .gitattributes file for the Godot Asset Library

nagidev committed:
Created FUNDING.yml for Ko-fi donations

nagidev committed:
Code refactor for undo/redo and bugfixes

nagidev committed:
Updated README.md to remove fixed issues

nagidev committed:
Fixed side panel toggle button

nagidev committed:
Updated README.md

nagidev committed:
Added DialogueParser & DialogueBubble nodes

nagidev committed:
Fixed docs and added DialogueBubble example

nagidev committed:
Fixed DialogueBubble crashing without Camera2D

nagidev committed:
Fixed DialogueBox not respecting option settings

Deccir committed:
Added csv translation logic and an example

Deccir committed:
Fixed character names not being translated after switching locale

Deccir committed:
fix spelling

u1over commented:
How do you export csv file to translate it? I surely misunderstanding something. But i can not figure out how to use localization through this addon's .tres files

Deccir commented:
> How do you export csv file to translate it? I surely misunderstanding something. But i can not figure out how to use localization through this addon's .tres files

Currently you have to create a separate .tres file for each language and then add the .tres file under Project > Localization > Remaps. Then you can specify wich file should be used instead of this one for each locale.

The changes proposed in this PR give the user the ability to use godots other methods of translating e.g. using a translation.csv. Only if nagi decides to merge this PR into the main branch can you use it.

u1over commented:
Thanks for the answer! A lot. I downloaded your branch, have all the updates. Could you explain how should i put my dialogues data to the csv file?

Deccir commented:
> Thanks for the answer! A lot. I downloaded your branch, have all the updates. Could you explain how should i put my dialogues data to the csv file?

Just to make sure: Before you use my changes you should fully understand how godots csv translation works. Also you should understand how to use the current state of this addon.

After that you have to create a dialogue .tres you the way you would normally do but instead of actually putting the dialogue, names and options in the text- / option- and speaker inputs you put a custom key in there that you can then reference in the translation.csv. 

It's best if you just look at the example i added. For that you have to clone my repo, start the project and select the ExampleCSVTranslation from the dropdown. If you look at the corresponding .tres (best through the add ons dialogue editor) and the translation.csv it should be clear what is happening.

Anything beyond that is up to you.

u1over commented:
Got it. Thank you!

nagidev commented:
Hi!
Thanks for the PR! I tried to run the CSV example file from the dialogue editor and it doesn't seem to be working properly.
![image](https://github.com/nagidev/DialogueNodes/assets/38583283/1925c7ea-728e-497e-99e3-8ce11c8b66b9)
Am I missing something or is this a known issue?

Deccir committed:
translation now works in editor

Deccir commented:
I completely forgot to test the translation in the editor .... whoops my bad.
I updated the PR and now it's working (at least on my machine :D)
I added a new MenuButton which is used to select the locale. 
It is a bit ugly because godot doesn't initialize the TranslationServer when running a tool. Therefore when running the dialog in the editor i had to get the current locales Translation Object from the TranslationServer and then use it to manually get the translated string. 

.... and while writing this i just realized that it will throw errors when no translation files are present. I'm on it :D

Deccir committed:
fixed error when no translations are present

Deccir commented:
Now it should be working properly

Deccir committed:
fix possibly missing locales

nagidev commented on addons/dialogue_nodes/Editor.tscn:
```diff
@@ -106,6 +106,13 @@ keep_pressed_outside = true
+keep_pressed_outside = true
+text = "Locale"
+switch_on_hover = true
+
```
I think we shouldn't include the locale in menu like this. Maybe we can plan this for a future update. Please remove this for now.

nagidev commented on addons/dialogue_nodes/objects/DialogueParser.gd:
```diff
@@ -124,19 +124,32 @@ func _process_start(dict : Dictionary):
+				var translated = translation.get_message(input)
+				return str(translated)
+		else:
+			translate_function = func(input): return input
```
Let's keep the translations to only show up in-game and not in the dialogue editor for now.

nagidev commented on addons/dialogue_nodes/objects/Character.gd:
```diff
@@ -6,3 +6,7 @@ extends Resource
+
+var translated_name : String :
+	get:
+		return tr(name)
```
I think you should keep the `tr()` call in the DialogueParser script.

Deccir replied:
I could move that `tr()` to the DialogueParser but then we would still need the `translated_name` property (or we overwrite the `name` property but that just feels wrong) since the `dialogue_processed` signal emitted by the DialogueParser contains the Character object and not just the simple translated name. And it would have to be overwritten in every `_process_dialogue` (since theoretically the language could be changed which is why i can't just check if the property is set). 

I could also instead of modifying the Character object create a new copy of it in each call to `_process_dialogue` and only translate the copys name. (and then either overwrite the `name` or set the `translated_name` if we keep the latter.)
 
Alternatively i can call `tr()` when speaker_label.text is set to speaker.name in the DialogueBox / DialogueBubble.

Which do you prefer?

nagidev replied:
The last option seems great. We can leave the translation call to be handled by the DialogueBox/DialogueBubble nodes. Maybe even put the translation code for the dialogue and option text to DialogueBox/DialogueBubble nodes?

Deccir replied:
Ok got it, just one more question. Should the `_on_dialogue_processed` inside of the DialogueBox and DialogueBubble emit `dialogue_processed` with the untranslated or translated data or maybe even both?

nagidev commented on addons/dialogue_nodes/objects/DialogueBox.gd:
```diff
@@ -281,7 +281,13 @@ func _on_dialogue_processed(speaker : Variant, dialogue : String, options : Arra
+			# Godots TranslationServer won't translate in the editor therefore it has to be done manually
+			var translation : Translation = TranslationServer.get_translation_object(TranslationServer.get_locale())
+			var translated_name = translation.get_message(speaker.name) if translation else speaker.name
+			speaker_label.text = str(translated_name)
```
Let's show the tag name in editor and translated name in-game.

nagidev commented on addons/dialogue_nodes/objects/DialogueBubble.gd:
```diff
@@ -280,7 +280,13 @@ func _on_dialogue_processed(speaker : Variant, dialogue : String, options : Arra
+			# Godots TranslationServer won't translate in the editor therefore it has to be done manually
+			var translation : Translation = TranslationServer.get_translation_object(TranslationServer.get_locale())
+			var translated_name = translation.get_message(speaker.name) if translation else speaker.name
+			speaker_label.text = str(translated_name)
```
Same changes as the DialogueBox script.

nagidev commented on addons/dialogue_nodes/editor/editor.gd:
```diff
@@ -52,6 +56,55 @@ func run_tree(start_node_idx : int):
+	var locale = locale_menu.get_popup().get_item_text(locale_idx)
+	TranslationServer.set_locale(locale)
+
+
```
Please remove this code as well since we only want the translations to work in-game for now.

nagidev reviewed (changes requested):
The PR is great, but I think we should rely heavily on Godot's translation options and avoid adding another layer of options in the dialogue editor for translations. Please make the requested changes, so I can merge these into main.
Thanks!

Deccir committed:
removed translation in editor feature (handling of speaker.name translation is still to be decided)

Deccir committed:
moved translation to DialogueBubble and Box
